nullaway/src/main/java/com/uber/nullaway/GenericsCheck.java

```diff
@@ -0,0 +1,145 @@
+        // invalid
+        if (!hasNullableAnnotation) {
+          GenericsCheck.invalidInstantiationError(tree, state, currentState, config);
+        }
+      }
+      index++;
+    }
+  }
+
+  /** Generics checks for parameterized typed trees * */
+  @SuppressWarnings("ModifiedButNotUsed")
+  public static void checkInstantiationForParameterizedTypedTree(
+      ParameterizedTypeTree tree, VisitorState state, NullAway currentState, Config config) {
+    if (!config.isJSpecifyMode()) {
+      throw new IllegalStateException("should only check type instantiations in JSpecify mode");
+    }
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    HashSet<Integer> nullableTypeArguments = new HashSet<Integer>();
+    JCTree.JCTypeApply baseTypeTree = (JCTree.JCTypeApply) tree;
+    Type t = baseTypeTree.type;
+    Type.ClassType classTyp = (Type.ClassType) t;
+    if (classTyp.tsym.getMetadata() != null) {
+      List<Attribute.TypeCompound> baseTypeAttributes =
+          classTyp.tsym.getMetadata().getTypeAttributes();
+      for (int i = 0; i < baseTypeAttributes.size(); i++) {
+        nullableTypeArguments.add(baseTypeAttributes.get(i).position.parameter_index);
+      }
+    }
+    for (int i = 0; i < typeArguments.size(); i++) {
+      boolean hasNullableAnnotation = false;
+      if (typeArguments.get(i).getClass().equals(JCTree.JCTypeApply.class)) {
+        ParameterizedTypeTree parameterizedTypeTreeForTypeArgument =
+            (ParameterizedTypeTree) typeArguments.get(i);
+        Type argumentType = ASTHelpers.getType(parameterizedTypeTreeForTypeArgument);
+        if (argumentType != null
+            && argumentType.getTypeArguments() != null
+            && argumentType.getTypeArguments().length() > 0) { // Nested generics
+          checkInstantiationForParameterizedTypedTree(
+              parameterizedTypeTreeForTypeArgument, state, currentState, config);
+        }
+      }
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        for (JCTree.JCAnnotation annotation : annotatedType.getAnnotations()) {
+          Attribute.Compound attribute = annotation.attribute;
+          hasNullableAnnotation =
+              hasNullableAnnotation
+                  || attribute.toString().equals("@org.jspecify.nullness.Nullable");
+          if (hasNullableAnnotation) {
+            break;
+          }
+        }
+      }
+      if (hasNullableAnnotation) {
+        if (!nullableTypeArguments.contains(i)) {
+          GenericsCheck.invalidInstantiationError(
+              typeArguments.get(i), state, currentState, config);
+        }
+      }
+    }
+  }
+
+  private static void invalidInstantiationError(
+      Tree tree, VisitorState state, NullAway currentState, Config config) {
+    ErrorBuilder errorBuilder = new ErrorBuilder(config, "", ImmutableSet.of());
```

Comment:
Here I think we should use the `errorBuilder` inside the `NullAway` object instead of creating a new one.  I propose we add a new method `NullAway.reportError(...)` that takes in an `ErrorMessage` and a `VisitorState` and then does the `state.reportMatch()` call.  With that change, you won't need the `config` here anymore (since that is also stored inside the `NullAway` object)